tests/libvmi/vmi.go

```diff
@@ -0,0 +1,132 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package libvmi
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+type Option func(vmiOptions)
+
+const DefaultVmiName = "testvmi"
+
+type action func(vmi *kvirtv1.VirtualMachineInstance)
+type vmiOptions map[int]action
+
+const defaultTestGracePeriod int64 = 0
+
+const (
+	// Options Identifiers
+	optNone = iota
+
+	optContainerImageName
+	optEnableRng
+	optResourceMemory
+	optInterface
+	optNetwork
+	optTerminationGracePeriod
+
+	optCloudInitNoCloudUserData
+	optCloudInitNoCloudUserDataB64
+	optCloudInitNoCloudNetworkData
+	optCloudInitNoCloudNetworkDataB64
+)
+
+var defaultVmiOptions = vmiOptions{
+	optInterface: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Devices.Interfaces = append(
+			vmi.Spec.Domain.Devices.Interfaces, primaryInterfaceDevice(),
+		)
+	},
+	optNetwork: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Networks = append(vmi.Spec.Networks, *kvirtv1.DefaultPodNetwork())
+	},
+	optTerminationGracePeriod: func(vmi *kvirtv1.VirtualMachineInstance) {
+		t := defaultTestGracePeriod
+		vmi.Spec.TerminationGracePeriodSeconds = &t
+	},
+	optResourceMemory: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: resource.MustParse("8192Ki"),
+		}
+	},
+}
+
+// WithRng adds `rng` to the the vmi devices.
+func WithRng() Option {
+	return func(o vmiOptions) {
+		o[optEnableRng] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Devices.Rng = &kvirtv1.Rng{}
+		}
+	}
+}
+
+// WithResourceMemory specifies the vmi memory resource.
+func WithResourceMemory(value string) Option {
+	return func(o vmiOptions) {
+		o[optResourceMemory] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+				k8sv1.ResourceMemory: resource.MustParse(value),
+			}
+		}
+	}
+}
+
+// New instantiates a new VMI configuration,
+// building its properties based on the specified With* options.
+func New(namespace, name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
+	options := defaultVmiOptions
```

Comment:
So we have a set of `With*` functions that register a callback to vmiOptions and then we iterate these options and use registered callbacks.

Why do we need that intermediate options structure? Can't just `WithRng` return the callback directly?

Reply:
You have just pointed your comment on the reason.

It is needed in order to support defaults and the ability to overwrite them (e.g `ResourceMemory`).
It also provides the ability to have control over the options in case order or some special behavior is needed.  A previous version used it to handle the random name before creating the vmi, but it was dropped in favor of not having the option of non-random at all. But this mat change if it will be needed later on (I expect the scenario to occur, as testing that two vmi/s with the same name cannot co-exist is there somewhere).

Reply:
You can handle defaults in the `baseVmi` function if you really want some test defaults. It sounds redundant and implicit. There already are defaults implemented in KubeVirt. For memory, this depends on the selected system, so it would make sense to set the default based on that. Or just let users set it.

If you want two different names, you use two differnt `WithName("FOO")`. Correct me if I'm wrong or provide a different example of why we cannot do that.

You want to replace the current helper code with something new. I'm all for that. But this way of doing things looks like a regression from simplicity/readability perspective to me.

Reply:
> You can handle defaults in the `baseVmi` function if you really want some test defaults.

Not if I want to handle all options in the same place, which is the case now.
When you handle them in a different place with different mechanics, it becomes less obvious and harder
to detect. When there is an opertunity to do it in the sample place, I take it.
Also, overwriting an already defined option is not a good path in my view.

> It sounds redundant and implicit. There already are defaults implemented in KubeVirt. For memory, this depends on the selected system, so it would make sense to set the default based on that. Or just let users set it.

Tests defaults do not and usually are not related to production ones. The random name of a vmi is a good example of it.
I used the memory as the existing code logic, it was not my plan to change logic with this refactor. My plan is to make it easier to read, understand and use over the current implementation.
If one wants to optimize it further, it can be done and should be done with time.

> If you want two different names, you use two differnt `WithName("FOO")`. Correct me if I'm wrong or provide a different example of why we cannot do that.

The name is now always random and follows the existing logic.
If one wants a non random name in the future, there will be a need to process the name before creating the vmi.

> You want to replace the current helper code with something new. I'm all for that. But this way of doing things looks like a regression from simplicity/readability perspective to me.

If this is a regression compared to the current code, then indeed you should nack this change.
IMO this is an improvement to the existing code, otherwise I would not have suggested it.
I am not convinced this should be further simplified at the moment and if you or someone else think it should, I welcome you to post such an improvement now or in the future.

Reply:
I appreciate the effort, but I will not ack this. Despite the code is valid, it seems too complex to me. Writing tests should be as simple as possible and this code is raising the bar for people that would need to extend it.

I would be happy to hear from other reviewers what do they think.